### PR TITLE
docs: thorough godoc for StatefulApiCollectorForFinalizableEntity

### LINKED ISSUE #4608: [Doc][Framework] Add Godocs to api_collector_with_state.go
## Documentation Scope
_api_collector_with_state.go_ in backend

## Describe the Change
Add thorough, detailed Godocs to fields/functions that are missing docs. Additionally, for fields that are functions we should document the order in which the functions will get called at runtime, and explain what the relation between them is. For example, for "ResponseParser" you'd say this is called after "Query", or for "GetNextPagerCustomData" you'd say this is called after "ResponseParser" and so on.

### PR DESCRIPTION
**Pull request by klesh**
### Summary
Added thorough godoc for StatefulApiCollectorForFinalizableEntity

### Does this close any open issues?
Closes #4608

### CODE REVIEW

File: backend/helpers/pluginhelper/api/api_collector_with_state.go [hunk 1]
```diff
@@ -122,18 +122,25 @@ func (m *ApiCollectorStateManager) Execute() errors.Error {
...
 
 // NewStatefulApiCollectorForFinalizableEntity aims to add timeFilter/diffSync support for
-// APIs that do NOT support filtering data by updated date. However, it comes with the
+// APIs that do NOT support filtering data by the updated date. However, it comes with the
 // following constraints:
 //  1. The entity is a short-lived object or it is likely to be irrelevant
 //     a. ci/id pipelines are short-lived objects
 //     b. pull request might took a year to be closed or never, but it is likely irrelevant
-//  2. The entity must be Finalizable: when it is finalized, no modification forever
+//  2. The entity must be Finalizable: when it is finalized, no modification would be allowed ever since
```
Comment (keon94) [inline backend/helpers/pluginhelper/api/api_collector_with_state.go:130]: I'd reword this to: "The entity must be finalizable, meaning no future modifications will happen to it."

Comment (klesh) [reply backend/helpers/pluginhelper/api/api_collector_with_state.go]: Updated

File: backend/helpers/pluginhelper/api/api_collector_with_state.go [hunk 2]
```diff
@@ -258,27 +265,31 @@ type FinalizableApiCollectorArgs struct {
...
+	Concurrency           int                                                                                         // required for Undetermined Strategy, number of concurrent requests
+	GetNextPageCustomData func(prevReqData *RequestData, prevPageResponse *http.Response) (interface{}, errors.Error) // required for Sequential Strategy, to extract the next page cursor from the given response
+	GetTotalPages         func(res *http.Response, args *ApiCollectorArgs) (int, errors.Error)                        // required for Determined Strategy, to extract the total number of pages from the given response
 }
+
+// FinalizableApiCollectorDetailArgs is the arguments for the detail collector
 type FinalizableApiCollectorDetailArgs struct {
 	FinalizableApiCollectorCommonArgs
-	BuildInputIterator func() (Iterator, errors.Error)
+	BuildInputIterator func() (Iterator, errors.Error) // required, build an iterator which would iterate all unfinished records in the database so they can be recollected and updated
```
Comment (keon94) [inline backend/helpers/pluginhelper/api/api_collector_with_state.go:294]: Let's add an extra sentence to explain which function would be using the elements of the iterator.

File: backend/helpers/pluginhelper/api/api_collector_with_state.go [hunk 3]
```diff
@@ -258,27 +265,31 @@ type FinalizableApiCollectorArgs struct {
...
 type FinalizableApiCollectorListArgs struct {
-	// optional, leave it be `nil` if API supports filtering by created date (Don't forget to set the Query)
-	GetCreated func(item json.RawMessage) (time.Time, errors.Error)
 	FinalizableApiCollectorCommonArgs
-	Concurrency           int
-	PageSize              int
-	GetNextPageCustomData func(prevReqData *RequestData, prevPageResponse *http.Response) (interface{}, errors.Error)
-	// need to consider the data missing problem: what if new data gets created during collection?
-	// GetTotalPages         func(res *http.Response, args *ApiCollectorArgs) (int, errors.Error)
+	GetCreated            func(item json.RawMessage) (time.Time, errors.Error)                                        // optional, leave it be `nil` if API supports filtering by updated date (Don't forget to set the Query)
```
Comment (keon94) [inline backend/helpers/pluginhelper/api/api_collector_with_state.go:284]: Let's add the "item" is the Json is the same Json on the response body available to the ResponseParser function.

File: backend/helpers/pluginhelper/api/api_collector_with_state.go [hunk 4]
```diff
@@ -258,27 +265,31 @@ type FinalizableApiCollectorArgs struct {
...
 type FinalizableApiCollectorCommonArgs struct {
-	UrlTemplate     string `comment:"GoTemplate for API url"`
-	Query           func(reqData *RequestData, createdAfter *time.Time) (url.Values, errors.Error)
-	Header          func(reqData *RequestData, createdAfter *time.Time) (http.Header, errors.Error)
-	MinTickInterval *time.Duration
-	ResponseParser  func(res *http.Response) ([]json.RawMessage, errors.Error)
-	AfterResponse   common.ApiClientAfterResponse
-	RequestBody     func(reqData *RequestData) map[string]interface{}
-	Method          string
+	UrlTemplate     string                                                                          // required, url path template for the request, e.g. repos/{{ .Params.Name }}/pulls
```
Comment (keon94) [inline backend/helpers/pluginhelper/api/api_collector_with_state.go:271]: Let's explain what the allowed variables on the URL template are. For example, ".Params" corresponds to whatever gets set on RawDataSubtaskArgs. Similarly for ".Input", etc.

**Review by keon94 — DISMISSED** [review]

File: backend/helpers/pluginhelper/api/api_collector_with_state.go (hunk 2 above)
Comment (klesh) [reply backend/helpers/pluginhelper/api/api_collector_with_state.go:294]: What do you mean?

File: backend/helpers/pluginhelper/api/api_collector_with_state.go (hunk 3 above)
Comment (klesh) [reply backend/helpers/pluginhelper/api/api_collector_with_state.go:284]: No, it is not, ReponseParser receives the full response body, a `list` most of the time.
The GetCreated expects a `item` in the `list`

File: backend/helpers/pluginhelper/api/api_collector_with_state.go (hunk 4 above)
Comment (klesh) [reply backend/helpers/pluginhelper/api/api_collector_with_state.go:271]: The option will be passed to the `api_collector`, It should be explained in detail there instead

File: backend/helpers/pluginhelper/api/api_collector_with_state.go (hunk 2 above)
Comment (keon94) [reply backend/helpers/pluginhelper/api/api_collector_with_state.go:294]: This function returns an iterator, right? Where will the elements of the iterator be used?

**Review by keon94 — DISMISSED** [review]

**Comment by keon94** [comment]
LGTM

**Review by keon94 — APPROVED** [review]
